extensions/pkg/controller/healthcheck/worker/helpers.go

```diff
@@ -79,41 +48,140 @@ func CheckMachineDeployment(deployment *machinev1alpha1.MachineDeployment) error
+		if err := checkConditionStatus(condition, machinev1alpha1.ConditionTrue); err != nil {
 			return err
 		}
 	}
 
 	for _, falseConditionType := range falseMachineDeploymentConditionTypes {
-		conditionType := string(falseConditionType)
 		condition := getMachineDeploymentCondition(deployment.Status.Conditions, falseConditionType)
 		if condition == nil {
 			continue
 		}
-		if err := general.CheckConditionState(conditionType, string(corev1.ConditionFalse), string(condition.Status), condition.Reason, condition.Message); err != nil {
+		if err := checkConditionStatus(condition, machinev1alpha1.ConditionFalse); err != nil {
 			return err
 		}
 	}
 
 	return nil
 }
 
+func checkMachineDeploymentsHealthy(machineDeployments []machinev1alpha1.MachineDeployment) (bool, *string, error) {
+	for _, deployment := range machineDeployments {
+		for _, failedMachine := range deployment.Status.FailedMachines {
+			return false, pointer.StringPtr("MachineDeploymentHasFailedMachines"), fmt.Errorf("machine %s failed: %s", failedMachine.Name, failedMachine.LastOperation.Description)
+		}
+
+		if err := CheckMachineDeployment(&deployment); err != nil {
+			return false, pointer.StringPtr("MachineDeploymentUnhealthy"), fmt.Errorf("machine deployment %s in namespace %s is unhealthy: %v", deployment.Name, deployment.Namespace, err)
+		}
+	}
+
+	return true, nil, nil
+}
+
+func checkNodesScalingUp(machineList *machinev1alpha1.MachineList, readyNodes, desiredMachines int) (gardencorev1beta1.ConditionStatus, *string, error) {
+	if readyNodes == desiredMachines {
+		return gardencorev1beta1.ConditionTrue, nil, nil
+	}
+
+	if machineObjects := len(machineList.Items); machineObjects < desiredMachines {
+		return gardencorev1beta1.ConditionFalse, pointer.StringPtr("MissingMachines"), fmt.Errorf("not enough machine objects created yet (%d/%d)", machineObjects, desiredMachines)
+	}
+
+	var pendingMachines, errorneousMachines int
+	for _, machine := range machineList.Items {
+		switch machine.Status.CurrentStatus.Phase {
+		case machinev1alpha1.MachineRunning, machinev1alpha1.MachineAvailable:
+			// machine is already running fine
+			continue
+		case machinev1alpha1.MachinePending:
+			// machine is being created
+			pendingMachines++
+		default:
+			// undesired machine phase
+			errorneousMachines++
+		}
+	}
+
+	if errorneousMachines > 0 {
+		return gardencorev1beta1.ConditionFalse, pointer.StringPtr("ErrorneousMachines"), fmt.Errorf("%d machine objects are errorneous", errorneousMachines)
+	}
+	if pendingMachines == 0 {
+		return gardencorev1beta1.ConditionFalse, pointer.StringPtr("MissingNodes"), fmt.Errorf("not enough ready worker nodes registered in the cluster (%d/%d)", readyNodes, desiredMachines)
+	}
+
+	return gardencorev1beta1.ConditionProgressing, pointer.StringPtr("PendingMachines"), fmt.Errorf("%d machines are being provisioned and should join the cluster and become ready soon", pendingMachines)
```

Comment:
The message could be improved for the case that there is only one machine.

Reply:
done